src/rgw/store/sfs/types.h

```diff
@@ -167,6 +341,55 @@ class Bucket {
 
   void delete_object(ObjectRef objref, const rgw_obj_key & key);
 
+  MultipartUploadRef get_multipart(
+    const std::string &upload_id,
+    const std::string &oid,
+    ACLOwner owner,
+    ceph::real_time mtime
+  ) {
+    std::lock_guard l(multipart_map_lock);
+
+    auto it = multiparts.find(upload_id);
+    if (it != multiparts.end()) {
+      auto mp = it->second;
+      ceph_assert(mp->objref);
+      ceph_assert(mp->objref->name == oid);
+      return mp;
+    }
+
+    ObjectRef obj = std::make_shared<Object>(oid);
+    MultipartUploadRef mp =
+      std::make_shared<MultipartUpload>(obj, upload_id, owner, mtime);
+    multiparts[upload_id] = mp;
+    return mp;
+  }
+
+  void finish_multipart(const std::string &upload_id, ObjectRef objref) {
+    std::lock_guard l(multipart_map_lock);
+
+    auto it = multiparts.find(upload_id);
+    ceph_assert(it != multiparts.end());
+    auto mp = it->second;
+    mp->finish();
+    multiparts.erase(it);
+
+    objref->metadata_finish(store);
+    finish(nullptr, objref->name);
+  }
+
+  std::string gen_multipart_upload_id() {
+    auto now = ceph::real_clock::now();
+    return ceph::to_iso_8601_no_separators(
+      now, ceph::iso_8601_format::YMDhmsn
+    );
+  }
+
+  std::map<std::string, MultipartUploadRef> get_multiparts() {
+    std::lock_guard l(multipart_map_lock);
+    // this should be doing a copy
```

Comment:
it is, I think we can get rid of this comment

Reply:
we can, but I left it there for a reason: to be clear we are doing a copy of the map, should someone less familiar with the code (or with c++) reads that and thinks we are returning the same map.